app/lib/simulator/Simulator.coffee

```diff
@@ -392,20 +392,40 @@ module.exports = class Simulator extends CocoClass
 
   generateSpellsObject: ->
     spells = {}
+    promises = []
     for {hero, team} in [{hero: 'Hero Placeholder', team: 'humans'}, {hero: 'Hero Placeholder 1', team: 'ogres'}]
       sessionInfo = _.filter(@task.getSessions(), {team: team})[0]
       fullSpellName = _.string.slugify(hero) + '/plan'
       submittedCodeLanguage = sessionInfo?.submittedCodeLanguage ? 'javascript'
       submittedCodeLanguage = 'javascript' if submittedCodeLanguage in ['clojure', 'io']  # No longer supported
       submittedCode = LZString.decompressFromUTF16 sessionInfo?.submittedCode?[_.string.slugify(hero)]?.plan ? ''
       aether = new Aether createAetherOptions functionName: 'plan', codeLanguage: submittedCodeLanguage, skipProtectAPI: false
-      try
-        aether.transpile submittedCode
-      catch e
-        console.log "Couldn't transpile #{fullSpellName}:\n#{submittedCode}\n", e
-        aether.transpile ''
-      spells[fullSpellName] = name: 'plan', team: team, thang: {thang: {id: hero}, aether: aether}
-    spells
+      spl = name: 'plan', team: team, thang: {thang: {id: hero}, aether: aether}, fullSpellName: fullSpellName
+      promises.push(@fetchToken(submittedCode, submittedCodeLanguage, spl))
+    Promise.all(promises)
+      .then((values) =>
+           for res in values
+             spl = res.spell
+             source = res.source
+             fullSpellName = spl.fullSpellName
+             try
+               spl.thang.aether.transpile source
+             catch e
+               console.log "Couldn't transpile #{fullSpellName}:\n#{source}\n", e
+             delete spl.fullSpellName
+             spells[fullSpellName] = spl)
+      .then(() => spells)
```

Comment:
need code suggest here. i don't know how to write a cleaner code without async/await in coffeescript@1

Reply:
I am concerned that this will not work unless the places that use this API handle a returned promise.  Even using async await, this change causes the return type of this method to change.
Instead of returning a spells object it is now returning a promise containing the spells object. Where is `generateSpellsObject` used? It will now have to handle waiting for the promise to resolve.

Reply:
oh good question! that's what i didn't notice. the `generateSpellsObject` is used by `god.createWorld` , whose code is these:
```js
  createWorld: ({spells, preload, realTime, justBegin, keyValueDb, synchronous}) ->
    console.log "#{@nick}: Let there be light upon #{@level.name}! (preload: #{preload})"
    userCodeMap = @getUserCodeMap spells
```
and the `getUserCodeMap` looks like won't wait for the promise:
```js
  getUserCodeMap: (spells) ->
    userCodeMap = {}
    for spellKey, spell of spells
      (userCodeMap[spell.thang.thang.id] ?= {})[spell.name] = spell.thang.aether.serialize()
    userCodeMap
```
bad things...  we need 'parse-code' request for cpp transpile, this is async, although i think ladder simulator can be run async but i'm not familiar with this logic :joy:

Reply:
have a quick look into simulation code, may be ...
```js

  commenceSimulationAndSetupCallback: ->
    @listenToOnce @god, 'infinite-loop', @onInfiniteLoop
    @listenToOnce @god, 'goals-calculated', @processResults
    @god.createWorld {spells: @generateSpellsObject()}

    # Search for leaks, headless-client only.
    # NOTE: Memwatch currently being ignored by Webpack, because it's only used by the server.
 ```
here if we can change the `@god.createWorld` to `@listenTo` **generateSpellsObjectFinish** event ?

Reply:
Also took a look and I really like `then(sp => god.createWorld {spells: sp} ` idea. I'd also put all the code beneath in the `then` method. Then all the code like `@memwatch` will correctly run immediately after the `createWorld` call.

I like this! Then all you need to do is `.then` chain off `commenceSingleSimulation` and `commenceSimulationAndSetupCallback`.

Then I don't think the `Spell.coffee` changes are required.

Reply:
The CPP token is treated as source code by esper. Thus once we have CPP token code the simulation should work the same as javascript or python.

Reply:
This looks so good!